src/cargo/ops/cargo_install.rs

```diff
@@ -316,40 +329,77 @@ fn install_one(
         );
     }
 
-    // Helper for --no-track flag to make sure it doesn't overwrite anything.
-    let no_track_duplicates = || -> CargoResult<BTreeMap<String, Option<PackageId>>> {
-        let duplicates: BTreeMap<String, Option<PackageId>> = exe_names(&pkg, &opts.filter)
-            .into_iter()
-            .filter(|name| dst.join(name).exists())
-            .map(|name| (name, None))
-            .collect();
-        if !force && !duplicates.is_empty() {
-            let mut msg: Vec<String> = duplicates
-                .iter()
-                .map(|(name, _)| format!("binary `{}` already exists in destination", name))
-                .collect();
-            msg.push("Add --force to overwrite".to_string());
-            bail!("{}", msg.join("\n"));
-        }
-        Ok(duplicates)
-    };
-
     // WARNING: no_track does not perform locking, so there is no protection
     // of concurrent installs.
     if no_track {
         // Check for conflicts.
-        no_track_duplicates()?;
+        no_track_duplicates(&pkg, opts, &dst, force)?;
     } else if is_installed(&pkg, config, opts, &rustc, &target, root, &dst, force)? {
         let msg = format!(
             "package `{}` is already installed, use --force to override",
             pkg
         );
         config.shell().status("Ignored", &msg)?;
-        return Ok(false);
+        return Ok(None);
     }
 
+    Ok(Some(pkg))
+}
+
+fn no_track_duplicates(
```

Comment:
this ends up being called once before compilation and once after compilation and I couldn't exactly figure out why. The one before makes sense (to short circuit compilation to avoid overwrites).

AFAICT, the call after compilation is simply to get a list of duplicates in the `force` case? Not sure if we expect the answer to be different the second time? Couldn't figure it out quickly, so kept the behavior the same.